src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentRequestSender.java

```diff
@@ -29,49 +35,156 @@
+        try {
+            final PaymentPortType paymentPort = adyenPaymentPortRegistry.getPaymentPort(countryIsoCode);
+            final T result = adyenCall.apply(paymentPort);
+            return new SuccessfulAdyenCall<T>(result);
+        } catch (final Exception e) {
+            logger.info("exception during adyen request", e);
+            return mapExceptionToCallResult(e);
+        }
+    }
+
+    /**
+     * Educated guess approach to transform CXF exceptions into error status codes.
+     * In the future if we encounter further different cases it makes sense to change this if/else structure to a map with lookup.
+     */
+    private <T> AdyenCallResult<T> mapExceptionToCallResult(final Exception e) {
+        //noinspection ThrowableResultOfMethodCallIgnored
+        final Throwable rootCause = Throwables.getRootCause(e);
+        if (rootCause instanceof SocketTimeoutException) {
+            // read timeout
+            if (rootCause.getMessage().contains("Read timed out")) {
+                return new UnSuccessfulAdyenCall<T>(RESPONSE_NOT_RECEIVED, e.getMessage());
+            } else if (rootCause.getMessage().contains("Unexpected end of file from server")) {
+                return new UnSuccessfulAdyenCall<T>(RESPONSE_INVALID, rootCause.getMessage());
+            }
+        } else if (rootCause instanceof SocketException) {
+            if (rootCause.getMessage().contains("Unexpected end of file from server")) {
+                return new UnSuccessfulAdyenCall<T>(RESPONSE_INVALID, e.getMessage());
+            }
+        } else if (rootCause instanceof HTTPException) {
+            // connection timeout
+            if (((HTTPException) rootCause).getResponseCode() == SERVICE_UNAVAILABLE) {
+                return new UnSuccessfulAdyenCall<T>(REQUEST_NOT_SEND, rootCause.getMessage());
+            }
+            // e.g. different response code or strange response
+            return new UnSuccessfulAdyenCall<T>(RESPONSE_INVALID, rootCause.getMessage());
+        } else if (rootCause instanceof SOAPFaultException) {
+            return new UnSuccessfulAdyenCall<T>(RESPONSE_INVALID, e.getMessage());
+        } else if (rootCause instanceof IOException) {
+            if (rootCause.getMessage().contains("Invalid Http response")) {
+                // unparsable data as response
+                return new UnSuccessfulAdyenCall<T>(RESPONSE_INVALID, e.getMessage());
+            } else if (rootCause.getMessage().contains("Bogus chunk size")) {
+                return new UnSuccessfulAdyenCall<T>(RESPONSE_INVALID, e.getMessage());
+            }
+        } else if (rootCause instanceof WstxEOFException) {
+            // happens for example when 301 with empty body is returned...
+            return new UnSuccessfulAdyenCall<T>(RESPONSE_INVALID, e.getMessage());
+        } else if (rootCause instanceof SoapFault) {
+            return new UnSuccessfulAdyenCall<T>(RESPONSE_ABOUT_INVALID_REQUEST, e.getMessage());
```

Comment:
> If that's the case, shouldn't it map to CANCELED instead (i.e. we can safely re-submit the payment after fixing the programming error)?

ah that's right. Thought CANCELED would only be for cases where our request is ok but did't went through (so it's repeatable without change). In case of CANCELED would KillBill do automatic retries (guess so)?

> We should double check (can be done in a separate PR).

added therefor some days ago https://jira.groupondev.com/browse/GPS-6153 . Should create issues here instead or?